### PR TITLE
Add contact code/topic validation

### LINKED ISSUE #4250: No validation in 'Add new contact' and 'Join to public chat' fields
### User Story

As a user, I want to get the validation error when adding an invalid key or public channel by mistake.

### Description
*Type*: Bug
*Summary*:  can insert and add invalid contact key; can insert and add invalid value to public chat; can click on 'Add' in field 'Join to public chat' with empty value.

#### Expected behavior
validation errors
3. Please enter a valid contact code
4. The topic name cannot be empty
5. A topic name can contain only following characters: a-z -

#### Actual behavior
3. The new contact is added
4. 
chat-null
5. New public chat is added

### Reproduction

1. Open StatusIm, log in
2. Click on +
3. Enter invalid contact key (i.e. '123') in 'Add new contact', click on 'Add'
4. Repeat 2,  click on 'Add' in 'Join to public chat' |(with empty value)
5. Repeat 2, enter invalid topic name (i. e. '123') in 'Join to public chat', click on 'Add'

### Additional Information
* Status version: StatusIm desktop (version 2018-05-07)
* Operating System: MacOS High Sierra 10.13.4, Ubuntu 18.04

### PR DESCRIPTION
**Pull request by vitvly**
Fix #4250, #4281, #4434 

This PR adds contact code/topic validation, and design changes to add contact screen and chat list. Also add contact and tab selection buttons have been updated to look according to the design.

There are some design issues:
  - some elements of design depend on ~~https://github.com/status-im/react-native-desktop/issues/255~~ (resolved with https://github.com/status-im/react-native-desktop/pull/261) and https://github.com/status-im/react-native-desktop/issues/264
  - we don't have a popup menu component yet (related to https://github.com/status-im/react-native-desktop/issues/256)
  - ~~according to https://zpl.io/aBElz8L, we also need to implement on-hover event handler (https://github.com/status-im/react-native-desktop/issues/257)~~ (decided to postpone it for now)

Design links: zpl.io/2ZgvMKw, https://zpl.io/aRqj4dv

### CODE REVIEW

File: src/status_im/ui/screens/add_new/new_public_chat/subs.cljs [hunk 1]
```diff
@@ -0,0 +1,12 @@
+(ns status-im.ui.screens.add-new.new-public-chat.subs
+  (:require [re-frame.core :as re-frame]
+            [status-im.ui.screens.add-new.new-public-chat.db :as db]
+            [cljs.spec.alpha :as spec]))
+
+(re-frame/reg-sub
+  :new-topic-error-message
```
Comment (rcullito) [inline src/status_im/ui/screens/add_new/new_public_chat/subs.cljs]: this is an interesting use of a subscription, to check a schema against the actual data :thinking: cool stuff!

File: src/status_im/ui/screens/desktop/main/add_new/views.cljs [hunk 2]
```diff
@@ -2,48 +2,74 @@
...
+         [react/text {:style {:font-size 16 :color (if chat-error "#939ba1" :white)}} "Start chat"]]]]]
+     ^{:key "choosecontact"}
+     [react/view
+      (when (seq contacts) [react/text {:style {:font-size 14}} "Or choose a contact"])
+      [react/view {:style {:margin-top 12}} 
+       (doall
+        (for [c contacts]
+        ^{:key (:whisper-identity c)}
+        [react/touchable-highlight {:on-press #(do
+                                                 (re-frame/dispatch [:set :contacts/new-identity (:whisper-identity c)])
```
Comment (rcullito) [inline src/status_im/ui/screens/desktop/main/add_new/views.cljs]: might this PR also fix #4276? these remind me of the updated handlers that Andrea was mentioning.

**Review by cammellos — APPROVED** [review]
We tend to keep styling in `styles.cljs` files, instead of inlining, but otherwise shared codebase looks ok

File: src/status_im/ui/screens/add_new/new_public_chat/subs.cljs [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+  :<- [:get :public-group-topic]
+  (fn [topic]
+    (when-not (spec/valid? ::db/topic topic)
+      "Group topic error")))
+
```
Comment (cammellos) [inline src/status_im/ui/screens/add_new/new_public_chat/subs.cljs]: extra line at the end-of-file

**Review by rcullito — APPROVED** [review]
nice PR vitaliy! changes look good. in response to your comment about the popups not working, were you trying with the one in `status-im.utils.utils` or somewhere else? That one was able to work locally for me but I'm not sure if that would address your needs in this PR.

File: src/status_im/ui/screens/desktop/main/add_new/views.cljs (hunk 2 above)
Comment (cammellos) [reply src/status_im/ui/screens/desktop/main/add_new/views.cljs]: yes that's one of the handlers you can use

Comment (rcullito) [reply src/status_im/ui/screens/desktop/main/add_new/views.cljs]: nice, thanks

File: src/status_im/ui/screens/add_new/new_public_chat/subs.cljs (hunk 1 above)
Comment (vitvly) [reply src/status_im/ui/screens/add_new/new_public_chat/subs.cljs]: @rcullito thanks for the pointer to `show-popup` in `utils.utils`! Unfortunately this is an alert-based popup, and we'd want to have multi-item clickable popups, more like proper context menus.

**Comment by vitvly** [comment]
Added separate styles ns.

File: src/status_im/ui/screens/add_new/new_public_chat/subs.cljs [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+  (:require [re-frame.core :as re-frame]
+            [status-im.ui.screens.add-new.new-public-chat.db :as db]
+            [cljs.spec.alpha :as spec]))
+
+(re-frame/reg-sub
+  :new-topic-error-message
+  :<- [:get :public-group-topic]
+  (fn [topic]
+    (when-not (spec/valid? ::db/topic topic)
+      "Group topic error")))
```
Comment (flexsurfer) [inline src/status_im/ui/screens/add_new/new_public_chat/subs.cljs]: i18n

File: src/status_im/ui/screens/desktop/main/add_new/styles.cljs [hunk 5]
```diff
@@ -0,0 +1,58 @@
+(ns status-im.ui.screens.desktop.main.add-new.styles)
+
```
Comment (flexsurfer) [inline src/status_im/ui/screens/desktop/main/add_new/styles.cljs]: wrong file formatting

File: src/status_im/ui/screens/desktop/main/add_new/views.cljs [hunk 6]
```diff
@@ -2,48 +2,72 @@
...
-    [react/view {:style {:flex 1 :background-color "#eef2f5"}}
-     ^{:key "newcontact"}
-     [react/view {:style {:height 64 :align-items :center :padding-horizontal 11 :justify-content :center}}
-      [react/text {:style {:font-size 16 :color :black :font-weight "600"}}
-       "Add new contact"]]
-     [react/view {:style {:height 1 :background-color "#e8ebec" :margin-horizontal 16}}]
-     [react/view {:style {:height     90 :margin-horizontal 16 :margin-bottom 16 :background-color :white :border-radius 12}}
-                          ;:box-shadow "0 0.5px 4.5px 0 rgba(0, 0, 0, 0.04)"}}
-      [react/view {:style {:flex-direction :row :margin-horizontal 16 :margin-top 16}}
+                  contacts      [:all-added-people-contacts]
```
Comment (flexsurfer) [inline src/status_im/ui/screens/desktop/main/add_new/views.cljs]: wrong formatting

File: src/status_im/ui/screens/desktop/main/add_new/views.cljs [hunk 7]
```diff
@@ -2,48 +2,72 @@
...
+                  chat-error [:new-contact-error-message]
+                  topic [:get :public-group-topic]
+                  topic-error [:new-topic-error-message]
+                  account [:get-current-account]]
+    [react/scroll-view
+     [react/view {:style styles/new-contact-view}
+      ^{:key "newcontact"}
+      [react/view {:style styles/new-contact-title}
+       [react/text {:style styles/new-contact-title-text}
+        "New Chat"]]
```
Comment (flexsurfer) [inline src/status_im/ui/screens/desktop/main/add_new/views.cljs]: i18n

**Review by flexsurfer — CHANGES_REQUESTED** [review]
please use our clojure style guide , for example maps values should be aligned , etc, all styles maps should be moved to styles ns, use i18n for labels

File: src/status_im/ui/screens/desktop/main/add_new/views.cljs [hunk 8]
```diff
@@ -70,4 +75,16 @@
        [react/touchable-highlight {:on-press #(when-not topic-error (re-frame/dispatch [:create-new-public-chat topic]))}
         [react/view {:style (merge styles/add-contact-button {:background-color (if topic-error "#eef2f5" "#4360df") })}
 
-         [react/text {:style (merge styles/add-contact-button-text {:color (if topic-error "#939ba1" :white)})} "Join public chat"]]]]]]))
+         [react/text {:style (merge styles/add-contact-button-text {:color (if topic-error "#939ba1" :white)})} "Join public chat"]]]]
```
Comment (vkjr) [inline src/status_im/ui/screens/desktop/main/add_new/views.cljs]: `topic-error` here and above can be passed as an argument to style to make all color decision in styles.

**Review by vkjr — APPROVED** [review]

**Comment by churik** [comment]
- [ ]  1) No "#" in "Join public chat" 
expected (green) - actual (red):
differenc
If you start typing on the field "Join public chat" focus should appear right after "#" char.
Now "#" is displayed only as a placeholder

- [x]  2) Public chats in chat lists look different in mobile and desktop versions
mobile (green) -  desktop (red):
kdfsksdf
Expected - like in mobile(color icons with #, but this can be subject to discuss).

- [x]  3) "Join public chat" or "Start chat" buttons are active after you already joined the chat
**Steps:**
- click on +
- select "status" from list (or insert valid contact code)
- click on + again
**Expected result:** "Join public chat"( or "Start chat") button is not active
**Actual result:** "Join public chat" (or "Start chat")button is active
**Video:** http://take.ms/yc1g6

- [ ]  4) Long messages/names should be displayed with ... ([issue in react-native](https://github.com/status-im/react-native-desktop/issues/264))
long

- [ ] 5) If time of the last message contains more 7 chars (i.e. '10:31 AM') it is overlapping with chat area 


- [x] 6) Chat list is unresponsive after several scrolls up/down (randomly)
**Prerequisites:**
1-1 chats, public chats are started
**Steps (not exact):**
- switch between chats 
- scroll up/down several times
**Expected result:** app is always responsive
**Actual result:** app is hanging
**Video:** http://take.ms/ItYiT

- [x] 7) "No messages yet" is missing
when chat is empty, 
expected (green) - actual (red):
no messages

**Comment by EugeOrtiz** [comment]
@siphiuel I can still see the old styles in the latest jenkins bundle https://jenkins.status.im/job/status-react/job/desktop/job/pull-requests/job/PR-4822/11/#
Should I download it from somewhere else?

<blockquote></blockquote>

**Comment by churik** [comment]
@EugeOrtiz yes, you should download it from particular PR build, in this case https://jenkins.status.im/job/status-react/job/desktop/job/pull-requests/view/change-requests/job/PR-4752/

<blockquote></blockquote>

File: src/status_im/ui/screens/desktop/main/add_new/views.cljs (hunk 8 above)
Comment (vitvly) [reply src/status_im/ui/screens/desktop/main/add_new/views.cljs]: Ah, great point, fixed!

**Comment by vitvly** [comment]
Used labels from i18n, aligned maps properly.

File: src/status_im/ui/components/desktop/tabs.cljs [hunk 9]
```diff
@@ -26,7 +27,7 @@
     [react/view {:style tabs.styles/tab-container}
      (let [icon (if active? icon-active icon-inactive)]
        [react/view
-        [icons/icon icon {:color (:color (tabs.styles/tab-icon active?))}]])
+        [icons/icon icon {:style {:tint-color (if active? "#4360df" "#6e777e")}}]])
```
Comment (vkjr) [inline src/status_im/ui/components/desktop/tabs.cljs]: @siphiuel, probably here we also should use colors namespace.

Comment (vitvly) [reply src/status_im/ui/components/desktop/tabs.cljs]: @vkjr I thought you wouldn't notice :) Fixed.

File: src/status_im/ui/screens/add_new/new_chat/db.cljs [hunk 10]
```diff
@@ -5,11 +5,12 @@
             [clojure.string :as string]))
 
 (defn- validate-pub-key [whisper-identity {:keys [address public-key]}]
```
Comment (flexsurfer) [inline src/status_im/ui/screens/add_new/new_chat/db.cljs:7]: i'm not sure about this change, how it affects mobile version

**Review by flexsurfer — APPROVED** [review]

**Review by rcullito — APPROVED** [review]

**Comment by EugeOrtiz** [comment]
List of UI/UX bug/fixes regarding this issue:

**GENERAL**
- [x] Disabled main button should not be clickable, or change color. **(High prio)**
- [ ] There is an extra grey line below the input fields titles. (Mid prio)
captura de pantalla 2018-06-21 a la s 9 28 15 a m

- [ ] Review color of “0x…” placeholder (it has low contrast now). (Mid prio)
- [ ] Enabled main button should have a hand cursor. (Mid prio)
- [ ] Distance between input field label and input field should be 8px. (Low prio)
- [ ] Distance between input field and contact list title should be 32px. (Low prio)
- [ ] Distance between contact list title and first contact avatar should be 12px. (Low prio)
- [ ] Avatar size in contact list should be 46x46. (Low prio)


**NEW 1:1 CHAT**
- [ ] Focus should be on input field when I choose “Start new chat” from + button when separate screens will be implemented. (Mid prio)

**NEW PUBLIC CHAT**
- [x] "#" symbol shouldn't disappear when I type the topic, it should be part of the topic (zpl.io/aBE71Xe)  **(High prio)** 
   - from @Chu: enabled as duplicate, it was reported as (1)
- [ ] If topic has an invalid character or capital letter, we should show an input field error explaining that there is an invalid character (zpl.io/aB5mnom, message TBD).  **(High prio)**
- [ ] Focus should be on input field when I choose “Join public chat” from + button when separate screens will be implemented. (Mid prio)
- [ ] Review length of input field (it’s too long now). (Mid prio)
- [ ] Main button needs more space to the sides (left and right padding). (Low prio)

**Comment by vitvly** [comment]
Pushed fixes for items #1,3,5,7. 'Add chat' buttons are not clickable now when invalid data is entered.

**Comment by vitvly** [comment]
#2 - changed to use color coming in chat item;
#4 - depends on status-im/react-native-desktop#264

**Comment by vitvly** [comment]
Could not reproduce #6 yet.

**Comment by asemiankevich** [comment]
@churik please review when you have time to

**Comment by churik** [comment]
Comment with las fixws/issues:
- [x]  1) No "#" in "Join public chat" 
If you start typing on the field "Join public chat" focus should appear right after "#" char.
Now "#" is displayed only as a placeholder. Not fixed.
Expected: zpl.io/aBE71Xe ('#' should be black and focus should be next to "#" char.)
Video: http://take.ms/KiTKl
- [x]  2) Public chats in chat lists look different in mobile and desktop versions
Consider as fixed, colors are different.
- [x]  3)  "Join public chat" or "Start chat" buttons are active after you already joined the chat - fixed
- [x]  4) Long messages/names should be displayed with ... (issue in react-native) - will create separate issue when PR is merged
- [ ]  5) If time of the last message contains more 7 chars (i.e. '10:31 AM') it is overlapping with chat area  - not fixed
Green (expected, mobile, IOS) - Red (Actual, desktop, Mac OSx)
expt

- [x]  6) Chat list is unresponsive after several scrolls up/down (randomly) - cannot reproduce as well because  of constant crash #5012, consider as resolved for now
- [x]   7) "No messages yet" is missing - fixed
- [x]   8) No unread badge at all - old one was removed, new is not added yet. If adding a badge is separate feature, let's display unread messages like previously in `master (just a number). 
- [x]   9) Custom name is not updated in chat list
             *Preconditions:* user A uses status-mobile and has custom username; User B uses status-desktop;
             *Steps:* 
                 - User B: tap on + and insert chat code of User A to "Contact code", click on "Start chat"
                 - User B: send message in chat with user A
                 - User A: add user B as contact
                 - User B: check name of user A in chat list and contact list
             *Expected result:*  custom username
             *Actual result:* [video](https://monosnap.com/file/tTEor3YJijMXBOCfEkMpdKPOo4oef4)
wrg

- [ ]   10) Public chat name is cut when it is selected
             *Steps:* 
                 - Join to any public chat (i.e. `#ethereum`)
                 - Click on it, pay attention to chat name
             *Expected result:* Can see full name if it is not overlapped with time (`#ethereum`)
             *Actual result:* last character is cut, [video]( https://monosnap.com/file/urIlw2eNPzWrXuWZCbuIPbtbkLRg2b)
wrchatname

- [ ]   11) List of preselected topics are different in mobile and in desktop
chann
- [x]   12) Wrong date is displayed for last messages if you scroll up and down in chat with 2+ days history 
             *Preconditions:* user joined to chat with 2+ days history (i.e. `#ethereum`)
             *Steps:* 
                 - scroll up to fetch all history 
                 - scroll down to last message
             *Expected result:* last messages are displaying below "Today"
             *Actual result:* [video](https://monosnap.com/file/YWGUazebZtMOjOSsAZpyxW1UpvNHF1
)
hjdshgfhdsfhg
- [x]   13) Crash after switching to chat with (12) issue
             *Reproducibility:* 100%
             *Note:* it is not reproducible on `desktop` branch with same steps
             *Preconditions:* user has chat with reproduced 12 issue(i.e. #ethereum); user joined to other public chat (i.e. #one, #two)
             *Steps:* 
                 - switch between #ethereum and #one, #two several times
                 - switch to #ethereum
             *Expected result:* no crash
             *Actual result:* [video](https://monosnap.com/file/YWGUazebZtMOjOSsAZpyxW1UpvNHF1
)
             *Logs:* [StatusIm.log](https://github.com/status-im/status-react/files/2159467/StatusIm.log), 
[CrashLog.log](https://github.com/status-im/status-react/files/2159468/crash4752Automatic.log)

**Comment by vitvly** [comment]
#13 - the attached macOS crash log contains identical entries to https://github.com/status-im/status-react/issues/5012, namely:

```
Thread 0 Crashed:: CrBrowserMain  Dispatch queue: com.apple.main-thread
0   libsystem_platform.dylib      	0x00007fff6ac7e169 _platform_memmove$VARIANT$Haswell + 585
1   StatusIm                      	0x0000000108c228ce 0x108bbf000 + 407758
2   StatusIm                      	0x0000000108c225a9 YGNode::insertChild(YGNode*, unsigned int) + 41
3   StatusIm                      	0x0000000108c14c5d YGNodeInsertChild + 125
4   StatusIm                      	0x0000000108bead76 UIManager::manageChildren(int, QList<int> const&, QList<int> const&, QList<int> const&, QList<int> const&, QList<int> const&) + 630
5   StatusIm                      	0x0000000108c3e52d 0x108bbf000 + 521517
6   org.qt-project.QtCore         	0x000000010b84b9dc QMetaMethod::invoke(QObject*, Qt::ConnectionType, QGenericReturnArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument) const + 1308
...
```

Most likely they have the same root cause.

**Comment by maxhora** [comment]
@siphiuel mentioned crash hopefully should be resolved with https://github.com/status-im/react-native-desktop/pull/280 or at least there should become much more ui issues related to this (pretty core thing)

**Comment by vitvly** [comment]
@maxris `much fewer` you must have meant, not `much more` :)

**Comment by maxhora** [comment]
OMG, yes! :) Sorry

**Comment by vitvly** [comment]
Issues 1, 5, 8, 9 resolved.
Issue 10 - must be related to RN-desktop internals.
Issue 11 - list of default public chats has been updated in `develop` branch, after the recent merge of develop into desktop.

**Comment by churik** [comment]
- 5 still can be reproduced
monosnap 2018-07-05 20-26-42 (only with AM)

- 12, 13 are created separately (#5012, #5070)

- 14 Last messages time is not aligned
diff_2

So, only 2 small things and I'll move it to ready!

**Comment by vitvly** [comment]
Issues 5 and 14 are related to https://github.com/status-im/status-react/issues/5104

**Comment by vitvly** [comment]
Item 12 and corresponding issue https://github.com/status-im/status-react/issues/5070 fixed

**Comment by churik** [comment]
Known:
- Design issues moved to  https://github.com/status-im/status-react/issues/5118, #5117
- Crash: added details to #5012 
- RN-related: #5117, #5113, #5104 
- New Chat and Join public chat in separate popup: #4434

**Pull request merged**